### ISSUE TITLE
[patch] Performance of Conch/SSH

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10904]
|<img alt="cooz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| cooz reported|
|-|-|
|Trac ID|trac#1325|
|Type|enhancement|
|Created|2005-11-11 04:23:52Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1325 1325
type__enhancement enhancement
reporter__cooz cooz
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1131683032000000 1131683032000000
changetime__1143329887000000 1143329887000000
version__ 
owner__ 
cc__z3p cc__cooz
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="cooz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cooz commented|
|-|-|

```
#!html
<pre>
When I was profiling my code I noticed that most of time is spent on simple 
counter in twisted/conch/ssh/transport.py. So I rewrote _Counter-class. My 
version is about 5 times faster than orginal and data transfer rate is almost 
doubled. Maybe someone could check that this is not breaking anything?

--- transport.py    2005-11-10 13:26:01.826437000 +0200
+++ transport.py.new    2005-11-10 13:35:57.804232000 +0200
@@ -20,6 +20,7 @@
 import sha
 import zlib
 import math # for math.log
+import array

 # external library imports
 from Crypto import Util
@@ -727,17 +728,22 @@
         iv=iv[:mod.block_size]
         self.count = getMP('\xff\xff\xff\xff'+iv)[0]
         self.bs = mod.block_size
+        self.count = Util.number.long_to_bytes(self.count - 1)
+        self.count = '\x00'*(self.bs-len(self.count)) + self.count
+        self.count = array.array('c', self.count)
+        self.len = len(self.count) - 1
+
     def __call__(self):
-        ret = MP(self.count)[4:]
-        if ret[0]=='\x00':
-            ret=ret[1:]
-        if len(ret) &lt; self.bs:
-            ret = '\x00'*(self.bs-len(ret)) + ret
-        self.count += 1
-        if self.count == 2L ** self.bs:
-            self.count = 0
-        return ret
+        i = self.len
+        while i > -1:
+            self.count[i] = n = chr((ord(self.count[i]) + 1) % 256)
+            if n == '\x00':
+                i -= 1
+            else:
+                return self.count.tostring()

+        self.count = array.array('c', '\x00'*self.bs)
+        return self.count.tostring()

 def buffer_dump(b, title = None):
     r = title or ''
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16447]) Apply conch optimization patch from cooz

Author: cooz
Reviewers: exarkun, itamar
Fixes [#10904](https://github.com/twisted/twisted/issues/10904)

This speeds up a conch-internal counter class making conch at least 
twice (perhaps as much as five times) as fast.

This also adds a unit test and some docstrings for that class.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
